### ISSUE TITLE
Pledge to drop support for python 2

### ISSUE THREAD

**Issue by mpharrigan** [msmbuilder/msmbuilder#852]
http://python3statement.github.io/

We should sign this pledge!

**Comment by cxhernandez** [comment]
+1

**Comment by mpharrigan** [comment]
I'd actually like to do this much sooner. Maybe the release-after-next (i.e. 3.7). comments?

**Comment by rmcgibbo** [comment]
-1 on any pledge.

**Comment by jchodera** [comment]
Why drop py27 support sooner?

**Comment by msultan** [comment]
I propose v4.0+ given that we are doing version releases every quarter. That should be 15ish months from now. That should hopefully allow for a smoother transition.

**Comment by cxhernandez** [comment]
I think we cater to a small enough community that signing this pledge is okay.... Also, the sooner we sign the more publicity we get (who knows how many names will be on that list in a year)

**Comment by mpharrigan** [comment]
> Why drop py27 support sooner?

Like the site says, supporting multiple versions of python "is a small but constant friction in the development of a lot of code". I haven't used python 2 other than to test for compatibility issues in the past two years. I don't know about the other @msmbuilder/developers but I think a lot of them use py3 primarily as well. I don't want subtle bugs to sneak in under py2 that we don't catch because no one uses it, while also promising users that we support py2.

A bigger reason why I want to drop py2 is to give back to the python community. The fractionization between 2 and 3 is not good for anyone. I fully blame the python developers for making choices that guaranteed this transition would be terrible, but I also think it would be foolish to go back to python 2 and eventually people will move to python 3. Anything I can do to speed that process up, I'm all for.

Minor reason 1 is: compatibility checks and `six` reduce the readability of the code. Python's greatest strength (imo) is readability. MSMBuilder code is read more than written

Minor reasons 2: there are some features that I've wished I could use. Automatic new-style classes, `TemporaryDirectory` as a context manager, `os.makedirs(exist_ok=True)`, subprocess improvements (`run` and `subprocess.DEVNULL`) just off the top of my head. There's also the matrix multiply operator, type hinting, and coroutines (although I don't think we're in any hurry to use those particular things)

**Comment by kyleabeauchamp** [comment]
There's a lot of legacy python applications that are not going to be migrated to py3 on the year timescale.  I think we should ask ourselves "will a sizable fraction of our users be negatively impacted by dropping py27"?  I'd only be willing to sign a pledge once the answer to that question is no.  

My life involves both py27 and py3 because I have a number of legacy tools that are not scheduled for py3k support.  IMHO, user convenience is basically a function of the least common denominator.  

Finally, don't we basically pledge API consistency with sklearn?  If so, then shouldn't we base our timeline on theirs?

**Comment by kyleabeauchamp** [comment]
My gut feeling is not in 2016 or 2017.  Maybe 2018?  I do see the advantage in signing the pledge with a far-off date as a way to "remind" others to get things moving.

**Comment by jchodera** [comment]
We're mostly py27. Dropping support for py27 early will hurt us. We are working on migrating everyone, but it's a lot of worth to switching everything. A lot of those tasks fall to me to main the lab infrastructure, so it personally takes time away from other more pressing things.

I'm pretty confident I can get everything switched over without trouble by 2018. But by the end of 2016? That would be really hard.

We certainly appreciate all the hard work you guys put into msmbuilder and also advocating what's good for the community, but this would effectively be a great way to alienate a bunch of your primary users and drive people straight to your competitor code (pyemma).

**Comment by kyleabeauchamp** [comment]
There are also several infrastructure projects that have not yet been ported.  See, e.g., http://lists.openstack.org/pipermail/openstack-dev/2016-June/097930.html

**Comment by mpharrigan** [comment]
> Finally, don't we basically pledge API consistency with sklearn? If so, then shouldn't we base our timeline on theirs?

No, we use the same API design as them and wrap some of their code. In fact, we'd "need" to be py3 only before they can drop support for py2 because we depend on them!

I see msmbuilder as a "leaf" in the dependency tree of the scientific python ecosystem. That's why I think it would be a great candidate for an early py3 migration. Porting the scripts that use msmbuilder would be a matter of changing a couple `xrange`s to `range`

@jchodera : point taken. I have some comments though. With `conda`, you can easily switch between python environments. It would be way too much work to re-write everything for py3. But it would be a relatively small amount of work to mandate that any *new* code be written for python3. To use old scripts is just a matter of switching to a different conda environment with the last version of msmbuilder to support python2. For reproducability, you should probably be using the same version of msmbuilder you used when initially writing the script anyways

**Comment by jchodera** [comment]
> I see msmbuilder as a "leaf" in the dependency tree of the scientific python ecosystem. 

That's basically saying "we don't want you using `msmbuilder` under the hood to build any tools". We're currently building tools that require MSM construction "under the hood", so this essentially translates to "we don't want you using msmbuilder for that".

The py2/py3 environment thing is certainly valuable, but we can't use py2 and py3 tools as dependencies for the *same* tool!

**Comment by mpharrigan** [comment]
It's hard to get a sense of the community using msmbuilder. There's very little activity on the issue tracker / pull requests / mailing list outside of a couple pande group members.

If you're building tools that extend msmbuilder, i'd love to know about them. We can feature you in the readme or somewhere

**Comment by jchodera** [comment]
> It's hard to get a sense of the community using msmbuilder. There's very little activity on the issue tracker / pull requests / mailing list outside of a couple pande group members.

I've been wondering about whether we should think of some ways we can reach our audience. Having a mailing list people can sign up for on a tool-by-tool basis to get low-traffic email updates (maybe with a quick optional survey after the email is captured) could be one way to do this.

Another interesting way is to have tools send a bit of info back when people use the tool to capture some usage data, including an opt-out scheme for those who don't want to be monitored. I haven't found a good totally free tool for doing this that doesn't require a bit of engineering, but [instrumental](https://instrumentalapp.com/docs/python) looks like a super-convenient way to figure out what kinds of features people are using in the tool.

> If you're building tools that extend msmbuilder, i'd love to know about them. We can feature you in the readme or somewhere

Our stuff is all under development at this time, but roughly, this is:
* automated pipelines for processing Folding@home data to build MSMs
* adaptive sampling techniques
* combinations of MSMs with multistate transition interface sampling

We're more than happy to hit you up for linkbacks once any of these tools reach some basic level of maturity.

**Comment by mpharrigan** [comment]
> Another interesting way is to have tools send a bit of info back when people use the tool to capture some usage data, including an opt-out scheme for those who don't want to be monitored.

This is something that's crossed my mind. I can see ardent privacy advocates railing against it. In my opinion, it's the way ordinary users can "give back" to the developers for free software. Keep me posted on your searches for a nice package for implementing this

**Comment by jchodera** [comment]
Ardent privacy advocates an always pay attention to the release notes and disable it. I think that is eminently fair for free software that one presumably is electing to install. It might become more complex if it is a library that might be pulled in silently as a dependency, but maybe there is a way to emit a message during installation.

**Comment by cxhernandez** [comment]
>  maybe there is a way to emit a message during installation

+1

**Comment by jchodera** [comment]
We could use post-link scripts for that purpose, I bet:
http://conda.pydata.org/docs/building/build-scripts.html

**Comment by cxhernandez** [comment]
> post-link scripts for that purpose

How would we handle non-conda installs though? A separate prompt during `python setup.py install`?

**Comment by jchodera** [comment]
> How would we handle non-conda installs though? A separate prompt during python setup.py install?

Can't you just have `setup.py` emit a warning as well?

For a warning, something like this may work:
```
This free software is made possible through federal grant funding.
In order to ensure continued support of this project, this package will periodically send anonymous usage statistics that is critical to justifying continued funding.
To disable this, you can set the `OMNIA_USAGE_STATISTICS` environment variable to `False`, but please remember that doing so makes it more difficult for us to continue to provide high-quality scientific software.
```
